File: frontend/src/services/connectors.ts
/**
 * API client for the connector system.
 */

export interface CredentialField {
  name: string;
  label: string;
  secret: boolean;
  placeholder: string;
  field_type?: string;   // "text" | "select", defaults to "text"
  options?: string[];    // only for select type
}

export interface ConnectorParam {
  name: string;
  description: string;
  required: boolean;
  default: string | null;
}

export interface ConnectorAction {
  name: string;
  description: string;
  params: ConnectorParam[];
}

export interface ConnectorInfo {
  name: string;
  display_name: string;
  description: string;
  connector_type?: string;   // "action" | "service"
  auth_type?: string;        // "api_key" | "oauth2"  (defaults to "api_key")
  credential_fields: CredentialField[];
  actions: ConnectorAction[];
  enabled: boolean;
  configured: boolean;
}

export interface ConnectorInvokeResponse {
  success: boolean;
  result: Record<string, unknown>;
  error: string | null;
}

function authHeaders(): Record<string, string> {
  const token = localStorage.getItem('auth_token');
  return token ? { Authorization: `Bearer ${token}` } : {};
}

export async function listConnectors(): Promise<ConnectorInfo[]> {
  const resp = await fetch('/api/connectors', { headers: authHeaders() });
  if (!resp.ok) throw new Error(`Failed to load connectors (${resp.status})`);
  const data = await resp.json();
  return data.connectors;
}

export async function getConnectorConfig(
  connectorName: string
): Promise<Record<string, string>> {
  const resp = await fetch(`/api/connectors/${connectorName}/config`, {
    headers: authHeaders(),
  });
  if (!resp.ok) throw new Error(`Failed to load connector config (${resp.status})`);
  const data = await resp.json();
  return data.config;
}

export async function saveConnectorConfig(
  connectorName: string,
  config: Record<string, string>
): Promise<void> {
  const resp = await fetch(`/api/connectors/${connectorName}/config`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json', ...authHeaders() },
    body: JSON.stringify({ config }),
  });
  if (!resp.ok) throw new Error(`Failed to save connector config (${resp.status})`);
}

export async function getOAuthStatus(
  connectorName: string
): Promise<{ connected: boolean; expires_at: string; scope: string }> {
  const resp = await fetch(`/api/connectors/${connectorName}/oauth/status`, {
    headers: authHeaders(),
  });
  if (!resp.ok) throw new Error(`Failed to get OAuth status (${resp.status})`);
  return resp.json();
}

export async function revokeOAuth(connectorName: string): Promise<void> {
  const resp = await fetch(`/api/connectors/${connectorName}/oauth/revoke`, {
    method: 'POST',
    headers: authHeaders(),
  });
  if (!resp.ok) throw new Error(`Failed to revoke OAuth (${resp.status})`);
}

export async function invokeConnector(
  connectorName: string,
  action: string,
  params: Record<string, unknown>
): Promise<ConnectorInvokeResponse> {
  const resp = await fetch(`/api/connectors/${connectorName}/invoke`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...authHeaders() },
    body: JSON.stringify({ action, params }),
  });
  if (!resp.ok) throw new Error(`Connector invoke failed (${resp.status})`);
  return resp.json();
}

File: backend/src/models/connectors.py
"""Pydantic models for the connector system."""

from __future__ import annotations

from typing import Any, Optional
from pydantic import BaseModel, Field


class CredentialField(BaseModel):
    """A credential input field for a connector's config dialog."""
    name: str
    label: str
    secret: bool = True
    placeholder: str = ""
    field_type: str = "text"       # "text" | "select"
    options: list[str] = Field(default_factory=list)  # only used when field_type="select"


class ConnectorParam(BaseModel):
    """A single parameter for a connector action."""
    name: str
    description: str
    required: bool = False
    default: Optional[str] = None


class ConnectorAction(BaseModel):
    """A callable action exposed by a connector."""
    name: str
    description: str
    params: list[ConnectorParam] = Field(default_factory=list)


class ConnectorInfo(BaseModel):
    """Public info about a connector, including per-user state."""
    name: str
    display_name: str
    description: str
    credential_fields: list[CredentialField]
    actions: list[ConnectorAction]
    enabled: bool = False
    configured: bool = False


class ConnectorListResponse(BaseModel):
    connectors: list[ConnectorInfo] = Field(default_factory=list)


class ConnectorConfigUpdate(BaseModel):
    """Write one or more config keys for a connector."""
    config: dict[str, str] = Field(
        default_factory=dict,
        description="Map of config_key -> config_value. Set '__enabled' to 'true'/'false'.",
    )


class ConnectorInvokeRequest(BaseModel):
    action: str
    params: dict[str, Any] = Field(default_factory=dict)


class ConnectorInvokeResponse(BaseModel):
    success: bool
    result: dict[str, Any] = Field(default_factory=dict)
    error: Optional[str] = None
